### PR TITLE
Provide CA Certificates via cloud-init's ca_certs module

### PR DESCRIPTION
**Pull request by AlexVulaj**
- Bug

## What does this PR do? / Related Issues / Jira

Currently, the curl flag we're using to provide a ca cert will _only_ use the provided cert - we want to use both the supplied SSL cert and RHEL's OEM trust bundle. This change moves away from the curl flag and instead uses the `ca_certs` module of cloud-init.

Resolves [OSD-24353](https://issues.redhat.com//browse/OSD-24353)

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have tested the functionality against gcp / aws, it doesn't cause any regression
- [x] I have added execution results to the PR's readme

## How to test this PR locally / Special Instructions
See [OSD-24353](https://issues.redhat.com//browse/OSD-24353) for setting up a transparent proxy. I used https://gitlab.cee.redhat.com/travi/osd-proxy-infra for setting up a non-transparent proxy.

I ran these new changes against
- a transparent proxy
- a non-transparent proxy
- a split-cert transparent proxy

In all situations, I saw the expected results. Elaborating more on the split-cert situation, I set up a transparent proxy using mitmproxy with `--ignore-hosts registry.redhat.io:443` in the service file. When I ran network verifier without passing a capath, _all_ URLs failed *_EXCEPT_* registry.redhat.io

## Logs 

With this change, the userdata template expands to something like the following:
```
#cloud-config

ca_certs:
  trusted:
    - |-
      -----BEGIN CERTIFICATE-----
      THE CERTIFICATE
      -----END CERTIFICATE-----
runcmd:
  - systemctl mask --now serial-getty@ttyS0.service
  - dmesg -D
  - echo "NV_CURLJSON_BEGIN" >/dev/ttyS0
  - export http_proxy= https_proxy=
  - curl --capath /etc/pki/tls/certs/ --retry 3 --retry-connrefused -t B -Z -s -I -m ${TIMEOUT} -w "%{stderr}${LINE_PREFIX}%{json}\n" ${CURLOPT} ${URLS} --proto =http,https,telnet ${TLSDISABLED_URLS_RENDERED} 2>/dev/ttyS0
  - echo "NV_CURLJSON_END" >/dev/ttyS0
power_state:
  delay: 2
  mode: poweroff
  message: Auto-terminating instance due to timeout
  timeout: 300
```

### CODE REVIEW

File: pkg/verifier/aws/entry_point.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func (a *AwsVerifier) ValidateEgress(vei verifier.ValidateEgressInput) *output.O
 		"TIMEOUT":          vei.Timeout.String(),
 		"HTTP_PROXY":       vei.Proxy.HttpProxy,
 		"HTTPS_PROXY":      vei.Proxy.HttpsProxy,
-		"CACERT":           base64.StdEncoding.EncodeToString([]byte(vei.Proxy.Cacert)),
```
Comment (abyrne55) [inline pkg/verifier/aws/entry_point.go:168]: We'll need to add a change to `legacy.Probe()` as it's still expecting a b64-encoded string

File: pkg/probes/curl/curl_json.go [hunk 2]
```diff
@@ -104,17 +105,31 @@ func (clp Probe) GetExpandedUserData(userDataVariables map[string]string) (strin
...
+				},
+			},
+		}
+
+		cloudInitYamlBytes, cloudInitMarshalErr := yaml.Marshal(&cloudInit)
+		if cloudInitMarshalErr != nil {
+			return "", fmt.Errorf("unable to create cloud init config: %w", cloudInitMarshalErr)
+		}
+
+		userDataVariables["CACERT_RENDERED"] = strings.TrimSpace(string(cloudInitYamlBytes))
```
Comment (abyrne55) [inline pkg/probes/curl/curl_json.go:132]: Neat approach, and smart way of avoiding complicated YAML escaping

**Review by abyrne55 — CHANGES_REQUESTED** [review]

File: pkg/verifier/aws/entry_point.go (hunk 1 above)
Comment (AlexVulaj) [reply pkg/verifier/aws/entry_point.go:168]: Good catch here. Looking at the `userdata-template.yaml` for the legacy probe, I see this line:
```
echo "${CACERT}" | base64 --decode > /proxy.pem
```
What do you think about just changing that to
```
echo "${CACERT}" > /proxy.pem
```

File: pkg/probes/legacy/userdata-template.yaml [hunk 3]
```diff
@@ -15,7 +15,7 @@ write_files:
       fi
       echo "Using IMAGE: $IMAGE" >> /var/log/userdata-output
       if [[ "${CACERT}" != "" ]]; then
-        echo "${CACERT}" | base64 --decode > /proxy.pem
+        echo "${CACERT}" > /proxy.pem
```
Comment (AlexVulaj) [inline pkg/probes/legacy/userdata-template.yaml:18]: @abyrne55 from your other comment - this should take care of things.

Comment (abyrne55) [reply pkg/probes/legacy/userdata-template.yaml:18]: This works as long as there's no double-quotes (or other surprise characters) in $CACERT... which is a bit big of an assumption for a user-provided input. I'd suggest reversing this change and instead encoding the $CACERT input into base64 within this probe's `GetExpandedUserData()`, or adding some input sanitization logic. Clunky, I know, but the goal of the legacy probe is to emulate the verifier's old behavior as closely as possible — which sometimes means holding our noses.

File: pkg/verifier/aws/entry_point.go (hunk 1 above)
Comment (abyrne55) [reply pkg/verifier/aws/entry_point.go:168]: See [my above reply](https://github.com/openshift/osd-network-verifier/pull/262/files#r1704582179)

File: pkg/probes/legacy/userdata-template.yaml (hunk 3 above)
Comment (AlexVulaj) [reply pkg/probes/legacy/userdata-template.yaml:18]: So to be clear - we'll keep the encoding in `entry_point` but just have the new curl probe decode it back? I'm cool with that!

File: pkg/probes/curl/userdata-template.yaml [hunk 4]
```diff
@@ -6,7 +6,7 @@ runcmd:
   - dmesg -D
   - echo "${USERDATA_BEGIN}" >/dev/ttyS0
   - export http_proxy=${HTTP_PROXY} https_proxy=${HTTPS_PROXY}
-  - curl --retry 3 --retry-connrefused -t B -Z -s -I -m ${TIMEOUT} -w "%{stderr}${LINE_PREFIX}%{json}\n" ${CURLOPT} ${URLS} --proto =http,https,telnet ${TLSDISABLED_URLS_RENDERED} 2>/dev/ttyS0
+  - curl --capath /etc/pki/tls/certs/ --retry 3 --retry-connrefused -t B -Z -s -I -m ${TIMEOUT} -w "%{stderr}${LINE_PREFIX}%{json}\n" ${CURLOPT} ${URLS} --proto =http,https,telnet ${TLSDISABLED_URLS_RENDERED} 2>/dev/ttyS0
```
Comment (abyrne55) [inline pkg/probes/curl/userdata-template.yaml:9]:
```suggestion
  - curl --capath /etc/pki/tls/certs/ --proxy-capath /etc/pki/tls/certs/ --retry 3 --retry-connrefused -t B -Z -s -I -m ${TIMEOUT} -w "%{stderr}${LINE_PREFIX}%{json}\n" ${CURLOPT} ${URLS} --proto =http,https,telnet ${TLSDISABLED_URLS_RENDERED} 2>/dev/ttyS0
```
Necessary for proper operation behind non-transparent forward proxies

**Review by abyrne55 — CHANGES_REQUESTED** [review]
Very good stuff. I re-ran tests behind a transparent proxy to great success. I was then able to tweak my terraform testing scripts enough to get a non-transparent proxy working, and that's when I realized that curl does some [extra SSL certificate checks](https://curl.se/docs/manpage.html#--proxy-capath) when non-transparent proxies are specified (e.g., via `osd-network-verifier egress --https-proxy="https://...`). So the suggestion below just tells curl to check the same trust store dir for proxy SSL certs as it does for "unproxied" (or transparently proxied) SSL certs.

**Comment by abyrne55** [comment]
Tested locally. Nice work!

/lgtm
/approve

**Pull request merged**